simulator/CMakeLists.txt

```diff
@@ -0,0 +1,147 @@
+#! cmake
+#! Building the scalar MIPS CPU simulator
+#! Soshin Konstantin
+#! soshinkonstantinv@gmail.com
+#! https://github.com/SoshinK
+
+cmake_minimum_required(VERSION 2.6)
+project(mipt-mips)
+
+set(CPPS infra/macro_test.cpp
+    infra/elf_parser/elf_parser.cpp 
+    infra/memory/memory.cpp 
+    infra/config/config.cpp 
+    infra/ports/ports.cpp 
+    infra/cache/cache_tag_array.cpp 
+    bypass/data_bypass.cpp 
+    fetch/fetch.cpp 
+    mips/mips_instr.cpp  
+    func_sim/func_sim.cpp 
+    core/perf_sim.cpp 
+    writeback/writeback.cpp 
+    simulator.cpp
+    mips/mips_register/mips_register.cpp
+    )
+
+set(TESTS 
+    infra/cache
+    infra/elf_parser
+    infra/memory
+    infra/config
+    infra/instrcache
+    infra/ports
+    infra/string
+    bpu
+    func_sim
+    core
+    mips/mips_register
+    )
+
+set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wall -Wextra -Wpedantic -Wold-style-cast -Werror")
+
+if(${DEBUG})
+    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -O0 -g")
+else()
+    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -O3")
+    set(CMAKE_LD_FLAGS "${CMAKE_CXX_FLAGS} -flto")
+endif()
+
+if(NOT ${RELEASE})
+    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -march=native")
+endif()
+
+if(APPLE) 
+    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++1z") 
+else() 
+    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++17")
+endif()
+
+add_executable(${PROJECT_NAME} main.cpp)
+
+foreach(ITER IN LISTS CPPS)
```

Comment:
Could you please add come comments to this place? It looks it generates library... but I'm not sure.

Reply:
Yes, it is generate OBJECT library: https://cmake.org/cmake/help/latest/command/add_library.html#id4
In fact, I am not sure that this will improve building

Reply:
I got this on my local machine:

```bash
-- The C compiler identification is GNU 4.8.4
-- The CXX compiler identification is GNU 7.2.0
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/g++-7
-- Check for working CXX compiler: /usr/bin/g++-7 -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
CMake Error at CMakeLists.txt:67 (target_sources):
  Unknown CMake command "target_sources".


-- Configuring incomplete, errors occurred!
```

Reply:
My CMake:

```bash
$> cmake --version
cmake version 2.8.12.2
```

Reply:
But Travis-CI is happy though.

Reply:
Oh Im sorry. I didnt notice that this command appeared in CMake 3.1. What should I do?

Reply:
MIPT-MIPS informal guideline is "we do not support old versions of tools", so CMake 3.1 would be one of prerequisites, please add this information to README.md as well.